File: LeetCode/2023-05/16_1335. 工作计划的最低难度.py
'''
Author: hakusai
Date: 2023-05-16 08:59:13
LastEditTime: 2023-05-16 09:56:04
'''
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache, reduce, cmp_to_key
from itertools import accumulate, combinations, permutations, product
from heapq import nsmallest, nlargest, heapify, heappop, heappush
from bisect import bisect_left, bisect_right
from math import factorial, gcd
from cmath import inf
from typing import List


class Solution:
    def minDifficulty(self, jobDifficulty: List[int], d: int) -> int:
        n = len(jobDifficulty)
        if n < d:
            return -1

        lru_cache(maxsize=None)
        def dfs(i, j):
            if i == 0:
                return max(jobDifficulty[:j+1])
            res, mx = inf, 0
            for k in range(j, i-1, -1):
                mx = max(mx, jobDifficulty[k])
                res = min(res, dfs(i-1, k-1)+mx)
            return res
        return dfs(d-1, n-1)

File: LeetCode/2023-05/16_1335. 工作计划的最低难度.java
/*
 * @Author: hakusai
 * @Date: 2023-05-16 08:59:37
 * @LastEditTime: 2023-05-16 21:24:15
 */
import java.util.*;
import java.io.*;
import java.math.*;

class Solution {
    private int[] a;
    private int[][] memo;

    public int minDifficulty(int[] a, int d) {
        int n = a.length;
        if (n < d) return -1;

        this.a = a;
        memo = new int[d][n];
        for (int i = 0; i < d; ++i)
            Arrays.fill(memo[i], -1); // -1 表示还没有计算过
        return dfs(d - 1, n - 1);
    }

    private int dfs(int i, int j) {
        if (memo[i][j] != -1) // 之前计算过了
            return memo[i][j];
        if (i == 0) { // 只有一天，必须完成所有工作
            int mx = 0;
            for (int k = 0; k <= j; k++)
                mx = Math.max(mx, a[k]);
            return memo[i][j] = mx;
        }
        int res = Integer.MAX_VALUE;
        int mx = 0;
        for (int k = j; k >= i; k--) {
            mx = Math.max(mx, a[k]); // 从 a[k] 到 a[j] 的最大值
            res = Math.min(res, dfs(i - 1, k - 1) + mx);
        }
        return memo[i][j] = res;
    }
}

